python/dgl/distributed/client.py

```
@@ -2,33 +2,155 @@
+    try:
+        # doesn't even have to be reachable
+        s.connect(('10.255.255.255', 1))
+        ip_addr = s.getsockname()[0]
+    except:
+        ip_addr = '127.0.0.1'
+    finally:
+        s.close()
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    s.bind(("", 0))
+    s.listen(1)
+    port = s.getsockname()[1]
+    s.close()
+
+    return ip_addr + ':' + str(port)
+
+def connect_to_server(ip_config, queue_size=20*1024*1024*1024, net_type='socket'):
     """Connect this client to server.
 
     Parameters
     ----------
-    server_namebook : ...
-        Server names.
+    ip_config : str
+        Path of server IP configuration file.
+    queue_size : int
+        Size (bytes) of client queue buffer (~20 GB on default).
+        Note that the 20 GB is just an upper-bound and DGL uses zero-copy and
+        it will not allocate 20GB memory at once.
+    net_type : str
+        networking type, e.g., 'socket' (on default) or 'mpi' (do not support yet).
 
     Raises
     ------
     ConnectionError : If anything wrong with the connection.
     """
-    pass
+    assert queue_size > 0, 'queue_size (%d) cannot be a negative number.' % queue_size
+    assert net_type in ('socket', 'mpi'), \
+    'net_type (%s) can only be \'socket\' or \'mpi\'.' % net_type
+    # Register some basic service
+    rpc.register_service(rpc.CLIENT_REGISTER,
+                         rpc.ClientRegisterReuqest,
+                         rpc.ClientRegisterResponse)
+    rpc.register_service(rpc.SHUT_DOWN_SERVER,
+                         rpc.ShutDownRequest,
+                         None)
+    server_namebook = rpc.read_ip_config(ip_config)
+    num_servers = len(server_namebook)
+    rpc.set_num_server(num_servers)
+    group_count = []
+    max_machine_id = 0
+    for server_info in server_namebook.values():
+        group_count.append(server_info[3])
+        if server_info[0] > max_machine_id:
+            max_machine_id = server_info[0]
```

what happens if the server ids given by a user aren't contiguous? should we throw an error? or it can be handled by the system?

---

The server id is created by read_ip_config() API. User will not set any ID of server or client.